File: cpp/20_valid_parentheses.cpp
/*
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true
Example 2:

Input: "()[]{}"
Output: true
Example 3:

Input: "(]"
Output: false
Example 4:

Input: "([)]"
Output: false
Example 5:

Input: "{[]}"
Output: true
*/

#include<iostream>
#include<string>
#include<stack>
#include<vector>

using namespace std;

bool isValid(string s) {
    if (s.empty()) {
        return true;
    }

    stack<char> st;
    for (auto& ch : s) {
        switch(ch) {
            case '{':
            case '[':
            case '(':
                st.push(ch);
                break;

            case '}':
                if (st.empty()) {
                    return false;
                }
                if (st.top() == '{') {
                    st.pop();
                } else {
                    st.push(ch);
                }
                break;

            case ']':
                if (st.empty()) {
                    return false;
                }
                if (st.top() == '[') {
                    st.pop();
                } else {
                    st.push(ch);
                }
                break;

            case ')':
                if (st.empty()) {
                    return false;
                }
                if (st.top() == '(') {
                    st.pop();
                } else {
                    st.push(ch);
                }
                break;

            default:
                return false;
        }
    }

    if (st.empty()) {
        return true;
    }

    return false;
}

void test() {
    string str1 = "(())";
    string str2 = "()[]{}";
    string str3 = "(]";
    string str4 = "([)]";
    string str5 = "{[]}";
    string str6 = "]";
    string str7 = "(])";
    vector<string> vstr;
    vstr.push_back(str1);
    vstr.push_back(str2);
    vstr.push_back(str3);
    vstr.push_back(str4);
    vstr.push_back(str5);
    vstr.push_back(str6);
    vstr.push_back(str7);

    for (auto& str : vstr) {
        bool valid = isValid(str);
        cout << "input: " << str;
        cout << (valid ? " is valid" : " is not valid") << endl;
    }
}

int main() {
    test();
    return 0;
}

File: py2/20_valid_parentheses.py
#!/bin/python2.7

"""
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true
Example 2:

Input: "()[]{}"
Output: true
Example 3:

Input: "(]"
Output: false
Example 4:

Input: "([)]"
Output: false
Example 5:

Input: "{[]}"
Output: true
"""


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for ch in s:
            if ch == "(" or ch == "[" or ch == "{":
                stack.append(ch)
                continue

            if len(stack)  == 0 and (ch == ")" or ch == "]" or ch == "}"):
                return  False

            if ch == ")"  and stack[-1] != "(":
                return  False

            if ch == "]"  and stack[-1] != "[":
                return  False

            if ch == "}"  and stack[-1] != "{":
                return  False

            stack.pop()

        if len(stack) == 0:
            return True

        return False

class Solution2(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """

        stack = []
        lookup = {")": "(", "]": "[", "}": "{"}
        for ch in s:
            if ch in lookup.values():
                stack.append(ch)
            elif ch in lookup.keys():
                if not stack or (stack and lookup[ch] != stack.pop()):
                    return False
            else:
                return False

        return not stack


def test():
    solu = Solution2()
    # s = "()"
    # s = "()[]{}"
    # s = "(]"
    # s = "([)]"
    s = "{[]}"
    result = solu.isValid(s)
    print("Output: " + str(result))


def main():
    test()


if __name__ == "__main__":
    main()
